pspmutating.go

```diff
@@ -0,0 +1,133 @@
+package pspmigrator
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-test/deep"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func GetContainerSecurityContexts(podSpec v1.PodSpec) []*v1.SecurityContext {
+	scs := make([]*v1.SecurityContext, 0)
+	for _, c := range podSpec.Containers {
+		scs = append(scs, c.SecurityContext)
+	}
+	return scs
+}
+
+func GetPSPAnnotations(annotations map[string]string) map[string]string {
+	pspAnnotations := make(map[string]string)
+	for ann, val := range annotations {
+		if strings.Contains(ann, "seccomp.security") || strings.Contains(ann, "apparmor.security") {
+			pspAnnotations[ann] = val
+		}
+	}
+	return pspAnnotations
+}
+
+// IsPodBeingMutatedByPSP returns whether a pod is likely mutated by a PSP object. It also returns the difference
+// of the securityContext attribute between the parent controller (e.g. Deployment) and the running pod.
+func IsPodBeingMutatedByPSP(pod *v1.Pod, clientset *kubernetes.Clientset) (mutating bool, diff []string, err error) {
+	diff = make([]string, 0)
+	if len(pod.ObjectMeta.OwnerReferences) > 0 {
+		var owner metav1.OwnerReference
+		for _, reference := range pod.ObjectMeta.OwnerReferences {
+			if reference.Controller != nil && *reference.Controller == true {
+				owner = reference
+				break
+			}
+		}
+		var parentPod v1.PodTemplateSpec
+		if owner.Kind == "ReplicaSet" {
+			rs, err := clientset.AppsV1().ReplicaSets(pod.Namespace).Get(context.TODO(), owner.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, diff, err
+			}
+			parentPod = rs.Spec.Template
+		}
+		if owner.Kind == "DaemonSet" {
+			ds, err := clientset.AppsV1().DaemonSets(pod.Namespace).Get(context.TODO(), owner.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, diff, err
+			}
+			parentPod = ds.Spec.Template
+		}
```

Comment:
Lots more controller types. See https://github.com/kubernetes/pod-security-admission/blob/master/admission/admission.go#L93. I guess a few of those are meta-controllers and can be ignored though. I'd consider structuring it something like this:

```go
import psaadmission "k8s.io/pod-security-admission/admission"

func FetchControllerPod(kind, name, namespace string) (*metav1.ObjectMeta, *corev1.PodSpec, error) {
	obj, err := fetchControllerObj(kind, name, namespace)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to fetch controller: %w", err)
	}
	return psaadmission.DefaultPodSpecExtractor.ExtractPodSpec(obj)
}

func fetchControllerObj(kind, name, namespace string) (runtime.Object, error) {
	switch kind {
	case "ReplicaSet":
		return clientset.AppsV1().ReplicaSets(namespace).Get(context.TODO(), name, metav1.GetOptions{})
	// Etc.
	default:
		return nil, fmt.Errorf("unsupported controller kind %s", kind)
	}
}
```

Reply:
Done. Much cleaner now. Thanks a lot.